src/coreclr/vm/appdomain.cpp

```diff
@@ -715,10 +715,10 @@ void BaseDomain::Init()
 #endif // FEATURE_COMINTEROP
 
     m_dwSizedRefHandles = 0;
-    if (!m_iNumberOfProcessors)
-    {
-        m_iNumberOfProcessors = GetCurrentProcessCpuCount();
-    }
+    // For server GC this value indicates the number of GC heaps used in circular order to allocate sized
+    // ref handles. It must not exceed the array size allocated by the handle table (see getNumberOfSlots
+    // in objecthandle.cpp). We might want to use GetNumberOfHeaps if it were accessible here.
+    m_iNumberOfProcessors = min(GetCurrentProcessCpuCount(), GetTotalProcessorCount());
```

Comment:
Note the comment in `getNumberOfSlots`: https://github.com/dotnet/runtime/blob/ebd695fee1ccccac472e39adb2a2469e89a7ccb2/src/coreclr/gc/objecthandle.cpp#L525-L536